File: connectors/connector-jira/src/test/scala/com/taskadapter/connector/jira/JiraTaskTypeResolverTest.scala
package com.taskadapter.connector.jira

import com.taskadapter.connector.definition.TaskId
import com.taskadapter.model.{GTask, GTaskBuilder, TaskType}
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.{FunSpec, Matchers}

@RunWith(classOf[JUnitRunner])
class JiraTaskTypeResolverTest extends FunSpec with Matchers {
  val someId = TaskId(1, "KEY1")
  val topLevel = "default parent"
  val sub = "default sub"
  val converter = GTaskToJiraFactory.getConverter()

  it("resolves default task type") {
    verify(GTaskBuilder.withSummary(), topLevel)
  }

  it("resolves default sub task type") {
    verify(GTaskBuilder.withSummary().setParentIdentity(someId), sub)
  }

  it("keeps explicit type") {
    verify(new GTask().setValue(TaskType, "Task"), "Task")
  }

  it("keeps explicit type even if parent is set") {
    verify(new GTask().setValue(TaskType, "Task").setParentIdentity(someId), "Task")
  }

  def verify(task: GTask, expected: String): Unit = {
    JiraTaskTypeResolver.resolveIssueTypeNameForCreate(converter.convertToJiraIssue(task),
      topLevel, sub) shouldBe expected
  }
}

File: connectors/connector-jira/src/test/java/com/taskadapter/connector/jira/JiraTaskTypeResolverTest.java
package com.taskadapter.connector.jira;

import com.taskadapter.connector.definition.TaskId;
import com.taskadapter.connector.definition.exception.FieldConversionException;
import com.taskadapter.model.AllFields;
import com.taskadapter.model.GTask;
import com.taskadapter.model.GTaskBuilder;
import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class JiraTaskTypeResolverTest {
    private final static TaskId someId = new TaskId(1L, "KEY1");
    private final static String topLevel = "default parent";
    private final static String sub = "default sub";
    private final static GTaskToJira converter = GTaskToJiraFactory.getConverter();

    @Test
    public void resolvesDefaultTaskType() throws FieldConversionException {
        verify(GTaskBuilder.withSummary(), topLevel);
    }

    @Test
    public void resolvesDefaultSubTaskType() throws FieldConversionException {
        verify(GTaskBuilder.withSummary().setParentIdentity(someId), sub);
    }

    @Test
    public void keepsExplicitType() throws FieldConversionException {
        verify(new GTask().setValue(AllFields.taskType, "Task"), "Task");
    }

    @Test
    public void keepsExplicitTypeEvenIfParentIsSet() throws FieldConversionException {
        verify(new GTask().setValue(AllFields.taskType, "Task").setParentIdentity(someId), "Task");
    }

    private static void verify(GTask task, String expected) throws FieldConversionException {
        assertThat(JiraTaskTypeResolver.resolveIssueTypeNameForCreate(converter.convertToJiraIssue(task),
                topLevel, sub)).isEqualTo(expected);
    }

}
